File: Patrones Estructurales/Bridge/bridge.py
from abc import ABC, abstractmethod

# 1. Implementor (Interfaz)
class Dispositivo(ABC):
    @abstractmethod
    def encender(self): pass
    @abstractmethod
    def set_volumen(self, percent): pass

# 2. Concrete Implementors
class TV(Dispositivo):
    def encender(self):
        print("TV encendida")
    def set_volumen(self, percent):
        print(f"TV volumen al {percent}%")

class Radio(Dispositivo):
    def encender(self):
        print("Radio encendida")
    def set_volumen(self, percent):
        print(f"Radio volumen al {percent}%")

# 3. Abstraction
class ControlRemoto:
    def __init__(self, dispositivo: Dispositivo):
        self.dispositivo = dispositivo # EL PUENTE

    def toggle_power(self):
        print("Presionando botón Power...")
        self.dispositivo.encender()

# 4. Refined Abstraction
class ControlAvanzado(ControlRemoto):
    def mute(self):
        print("Presionando botón Mute...")
        self.dispositivo.set_volumen(0)

# Uso
mi_radio = Radio()
control = ControlAvanzado(mi_radio)

control.toggle_power()
control.mute()

File: Patrones Estructurales/Bridge/bridge.js
// 1. Implementor (La Interfaz del dispositivo)
class Dispositivo {
    encender() { throw new Error("Método no implementado"); }
    apagar() { throw new Error("Método no implementado"); }
    setVolumen(percent) { throw new Error("Método no implementado"); }
}

// 2. Concrete Implementors (Dispositivos reales)
class TV extends Dispositivo {
    encender() { console.log("TV encendida"); }
    apagar() { console.log("TV apagada"); }
    setVolumen(percent) { console.log(`TV volumen al ${percent}%`); }
}

class Radio extends Dispositivo {
    encender() { console.log("Radio encendida"); }
    apagar() { console.log("Radio apagada"); }
    setVolumen(percent) { console.log(`Radio volumen al ${percent}%`); }
}

// 3. Abstraction (El control remoto básico)
class ControlRemoto {
    constructor(dispositivo) {
        this.dispositivo = dispositivo; // EL PUENTE (Bridge)
    }

    togglePower() {
        console.log("Presionando botón Power...");
        this.dispositivo.encender(); 
    }
}

// 4. Refined Abstraction (Un control más avanzado)
class ControlAvanzado extends ControlRemoto {
    mute() {
        console.log("Presionando botón Mute...");
        this.dispositivo.setVolumen(0);
    }
}

// Uso
const miTv = new TV();
const control = new ControlAvanzado(miTv);

control.togglePower(); // Usa la TV
control.mute();        // Usa la TV